src/main/java/org/logstash/filewatch/WatchedFilesCollection.java

```diff
@@ -0,0 +1,229 @@
+               .defineAnnotatedMethods(WatchedFilesCollection.class);
+    }
+
+    @JRubyMethod
+    public IRubyObject initialize(final ThreadContext context, IRubyObject settings) {
+        final String sort_by = settings.callMethod(context, "file_sort_by").asJavaString();
+        final String sort_direction = settings.callMethod(context, "file_sort_direction").asJavaString();
+
+        Comparator<IRubyObject> comparator;
+        switch (sort_by) {
+            case "last_modified" :
+                sortBy = "modified_at";
+                comparator = (file1, file2) -> {
+                    if (file1 == file2) return 0; // fast shortcut
+                    RubyFloat mtime1 = modified_at(context, file1);
+                    RubyFloat mtime2 = modified_at(context, file2);
+                    int cmp = Double.compare(mtime1.getDoubleValue(), mtime2.getDoubleValue());
+                    // if mtime same (rare unless file1 == file2) - order consistently
+                    if (cmp == 0) return path(context, file1).op_cmp(path(context, file2));
+                    return cmp;
+                };
+                break;
+            case "path" :
+                sortBy = "path";
+                comparator = (file1, file2) -> path(context, file1).op_cmp(path(context, file2));
+                break;
+            default :
+                throw context.runtime.newArgumentError("sort_by: '" + sort_by + "' not supported");
+        }
+
+        if ("desc".equals(sort_direction)) {
+            comparator = comparator.reversed();
+        }
+
+        this.files = new TreeMap<>(comparator);
+        this.filesInverse = RubyHash.newHash(context.runtime);
+
+        // variableTableStore("@files", JavaUtil.convertJavaToRuby(context.runtime, this.files));
+        // variableTableStore("@files_inverse", this.filesInverse);
+
+        return this;
+    }
+
+    @JRubyMethod
+    public IRubyObject add(ThreadContext context, IRubyObject file) {
+        RubyString path = getFilePath(context, file);
+        synchronized (this) {
+            RubyString prev_path = this.files.put(file, path);
+            assert prev_path == null || path.equals(prev_path); // file's path should not change!
+            this.filesInverse.op_aset(context, path, file);
+        }
+        return path;
+    }
+
+    private static RubyString getFilePath(ThreadContext context, IRubyObject file) {
+        IRubyObject path = file.callMethod(context, "path");
+        if (!(path instanceof RubyString)) {
+            throw context.runtime.newTypeError("expected file.path to return String but did not file: " + file.inspect());
+        }
+        if (!path.isFrozen()) path = ((RubyString) path).dupFrozen(); // path = path.dup.freeze
+        return (RubyString) path;
+    }
+
+    @JRubyMethod
+    public IRubyObject remove_paths(ThreadContext context, IRubyObject arg) {
```

Comment:
can this return `RubyFixnum` instead of `IRubyObject`?

Reply:
could but not much worth the effort - the ext caller piece won't care. 
would be worth the effort if we would need to call this from other Java parts ...